x-pack/plugins/fleet/server/services/register_integration_fields_extractor.ts

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { CoreSetup } from '@kbn/core/server';
+import type { FieldsMetadataServerSetup } from '@kbn/fields-metadata-plugin/server';
+
+import type { FleetStartContract, FleetStartDeps } from '../plugin';
+
+interface RegistrationDeps {
+  core: CoreSetup<FleetStartDeps, FleetStartContract>;
+  fieldsMetadata: FieldsMetadataServerSetup;
+}
+
+export const registerIntegrationFieldsExtractor = ({ core, fieldsMetadata }: RegistrationDeps) => {
+  fieldsMetadata.registerIntegrationFieldsExtractor(async ({ integration, dataset }) => {
+    const [_core, _startDeps, { packageService }] = await core.getStartServices();
+
+    return packageService.asInternalUser.getPackageFieldsMetadata({
+      packageName: integration,
```

Comment:
I'm trying to figure out how the name of being handled in this code. Is this exposed in any way? If so, I might be overlooking it. 

If code relies on this, how can we ensure that it fails quickly if the fleet plugin isn't available?

Reply:
I'm not sure about the initial question, the steps taken to expose a utility capable of getting integration fields in fleet were:
- expose a new method `getPackageFieldsMetadata` as part of the fleet internal `PackageService`.
- once the fleet plugin is initialize, register from the fieldsMetadata a callback to retrieve the fields given some parameters.

The reason behind this is that I don't want the fieldsMetadata plugin directly depending on fleet (and bring this dependency to all the consuming apps)n but inverting the dependency we delegate to fleet the job of registering this extractor.

> If code relies on this, how can we ensure that it fails quickly if the fleet plugin isn't available?

If fleet is not available because is disabled, the callback will not be registered in the plugin setup step, and the fieldsMetadata plugin will default to a placeholder function that simply does not return integration fields.

I don't think this should fail or return a failing response to the client as there might be many ECS fields in the requested selection that are always available from the default static ECS defs, a lack of permission on the integrations should fail if this was the primary resource, but as it's part of a field resolution step, the lack of access to fleet shouldn't affect the retrieval of the other fields.

Reply:
@tonyghiani How does one specify that they want this field info registry.

Reply:
Oh I think I see what you mean, otherwise correct me or let's chat more offline, happy to jump in a quick call.

Since this extractor have to fetch the field assets for a specific integration+dataset from EPR, with the performed request by `getPackageFieldsMetadata` we get access to the whole package assets anyway.

In this step we parse the fields for specified dataset (less assets to parse) or for all of them (using `dataset=*`) and we return a map of each available dataset fields:
```
// integration: nginx
// dataset: nginx.access

{
  "nginx.access": {
    "field1": {},
    "field2": {},
  }
}

// integration: nginx
// dataset: *

{
  "nginx.access": {
    "field1": {},
    "field2": {},
  },
  "nginx.errors": {
    "field1": {},
    "field2": {},
  }
}
```

This extractor registered with fleet has knowledge only about how to extract integration fields, not how to interact with them.

Once this extractor is registered, it is used by the `IntegrationFieldsRepository` as a dependency, where the methods to access specific fields access the returned results:
```ts
async getByName(
    fieldName: IntegrationFieldName,
    { integration, dataset }: IntegrationFieldsSearchParams
  ): Promise<FieldMetadata | undefined> {
    let field = this.getCachedField(fieldName, { integration, dataset });

    if (!field) {
      try {
        await this.extractFields({ integration, dataset });
      } catch (error) {
        throw new PackageNotFoundError(error.message);
      }

      field = this.getCachedField(fieldName, { integration, dataset });
    }

    return field;
  }
```

Please let me know if this answers your question or let me know if there is any example you had in mind for this.